File: app/src/main/java/com/gee12/mytetroid/activities/IconsActivity.java
package com.gee12.mytetroid.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.gee12.mytetroid.R;
import com.gee12.mytetroid.adapters.IconsListAdapter;
import com.gee12.mytetroid.data.DataManager;
import com.gee12.mytetroid.data.IconsManager;
import com.gee12.mytetroid.logs.TetroidLog;
import com.gee12.mytetroid.model.TetroidIcon;
import com.gee12.mytetroid.model.TetroidNode;

import java.io.File;
import java.util.List;
import java.util.OptionalInt;
import java.util.stream.IntStream;

/**
 * Активность для выбора иконки ветки.
 */
public class IconsActivity extends AppCompatActivity {

    public static final String EXTRA_NODE_ID = "EXTRA_NODE_ID";
    public static final String EXTRA_NODE_ICON_PATH = "EXTRA_NODE_ICON_PATH";
    public static final String EXTRA_IS_DROP = "EXTRA_IS_DROP";

    private Spinner spFolders;
    private ArrayAdapter<String> mFoldersAdapter;
    private IconsListAdapter mIconsAdapter;
    private View btnSubmit;
    private String mNodeId;
    private TetroidIcon mSelectedIcon;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_icons);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        // список иконок
        ListView mListView = findViewById(R.id.list_view_icons);
        this.mIconsAdapter = new IconsListAdapter(this, IconsManager.getIconsFolderPath());
        mListView.setAdapter(mIconsAdapter);
        mListView.setOnItemClickListener((adapterView, view, i, l) -> {
            TetroidIcon icon = (TetroidIcon) mIconsAdapter.getItem(i);
            if (icon != null) {
                this.mSelectedIcon = icon;
                mIconsAdapter.setSelectedIcon(icon);
                mIconsAdapter.setSelectedView(view);
                btnSubmit.setEnabled(true);
            }
        });

        // список каталогов
        this.spFolders = findViewById(R.id.spinner_folders);
        List<String> folders = IconsManager.getIconsFolders(this);
        if (folders != null) {
            if (folders.size() == 0) {
                TetroidLog.log(this, R.string.log_icons_dirs_absent, Toast.LENGTH_LONG);
                return;
            }
            this.mFoldersAdapter = new ArrayAdapter<>(this,
                    android.R.layout.simple_spinner_item, folders);
            mFoldersAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spFolders.setAdapter(mFoldersAdapter);

            spFolders.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    String folder = mFoldersAdapter.getItem(position);
                    if (folder != null) {
                        List<TetroidIcon> icons = IconsManager.getIconsFromFolder(IconsActivity.this, folder);
                        if (icons != null) {
                            // пришлось заново пересоздавать адаптер, т.к. при простой установке dataSet
                            // с помощью reset(icons)
                            // не сбрасывалось выделение от предыдущего списка icons (из другого раздела)
                            mIconsAdapter = new IconsListAdapter(IconsActivity.this, IconsManager.getIconsFolderPath());
                            mListView.setAdapter(mIconsAdapter);
                            if (mSelectedIcon != null && folder.equals(mSelectedIcon.getFolder())) {
                                String selectedIconName = mSelectedIcon.getName();
                                OptionalInt iconPos = IntStream.range(0, icons.size())
                                        .filter(i -> selectedIconName.equals(icons.get(i).getName()))
                                        .findFirst();
                                if (iconPos.isPresent()) {
                                    // выделяем существующую иконку
                                    int pos = iconPos.getAsInt();
                                    TetroidIcon icon = icons.get(pos);
                                    mSelectedIcon = icon;
                                    mIconsAdapter.setSelectedIcon(icon);
                                    btnSubmit.setEnabled(true);
                                }
                                mIconsAdapter.setData(icons);
                                if (iconPos.isPresent()) {
                                    int pos = iconPos.getAsInt();
                                    mListView.setSelection(pos);
                                }
                            } else {
//                                mIconsAdapter.setSelectedIcon(null);
//                                mIconsAdapter.setSelectedView(null);
                                mIconsAdapter.setData(icons);
//                                mListView.setItemChecked(-1, true);
//                                mListView.setSelection(-1);
//                                mListView.setSelectionAfterHeaderView();
                            }
                        }
                    }
                }
                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                }
            });
        } else {
            TetroidLog.log(this,
                    getString(R.string.log_icons_dir_absent_mask, DataManager.ICONS_FOLDER_NAME), Toast.LENGTH_LONG);
        }

        findViewById(R.id.button_clear_icon).setOnClickListener(view -> setResult(null));
        this.btnSubmit = findViewById(R.id.button_submit);
        btnSubmit.setOnClickListener(view -> setResult(mSelectedIcon));

        // получаем данные
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            mNodeId = extras.getString(EXTRA_NODE_ID);
            String iconPath = extras.getString(EXTRA_NODE_ICON_PATH);
            selectCurrentIcon(iconPath);
        }
    }

    private void selectCurrentIcon(String iconPath) {
        if (!TextUtils.isEmpty(iconPath)) {
            String[] pathParts = iconPath.split(File.separator);
            if (pathParts.length >= 2) {
                String name = pathParts[pathParts.length - 1];
                String folder = pathParts[pathParts.length - 2];
                mSelectedIcon = new TetroidIcon(folder, name);

                if (mFoldersAdapter != null) {
                    int pos = mFoldersAdapter.getPosition(folder);
                    if (pos >= 0) {
                        spFolders.setSelection(pos);
                    }
                }
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Respond to the action bar's Up/Home button
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setResult(TetroidIcon icon) {
        Intent intent = new Intent();
        intent.putExtra(EXTRA_NODE_ID, mNodeId);
        intent.putExtra(EXTRA_NODE_ICON_PATH, (icon != null) ? icon.getPath() : null);
        intent.putExtra(EXTRA_IS_DROP, icon == null);
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    public static void startIconsActivity(Activity activity, TetroidNode node, int requestCode) {
        if (node == null) {
            return;
        }
        Intent intent = new Intent(activity, IconsActivity.class);
        intent.putExtra(EXTRA_NODE_ID, node.getId());
        intent.putExtra(EXTRA_NODE_ICON_PATH, node.getIconName());
        activity.startActivityForResult(intent, requestCode);
    }
}

File: app/src/main/java/com/gee12/mytetroid/ui/node/icon/IconsActivity.kt
package com.gee12.mytetroid.ui.node.icon

import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.annotation.RequiresApi
import com.gee12.mytetroid.R
import com.gee12.mytetroid.common.Constants
import com.gee12.mytetroid.common.extensions.buildIntent
import com.gee12.mytetroid.di.ScopeSource
import com.gee12.mytetroid.model.TetroidIcon
import com.gee12.mytetroid.model.TetroidNode
import com.gee12.mytetroid.ui.base.TetroidStorageActivity
import com.gee12.mytetroid.ui.base.BaseEvent

/**
 * Активность для выбора иконки ветки.
 */
class IconsActivity : TetroidStorageActivity<IconsViewModel>() {

    private lateinit var spFolders: Spinner
    private lateinit var iconsListView: ListView
    private lateinit var foldersAdapter: ArrayAdapter<String>
    private lateinit var iconsAdapter: IconsListAdapter
    private lateinit var btnSubmit: Button
    private lateinit var btnClear: Button
    private var selectedIcon: TetroidIcon? = null


    override fun getLayoutResourceId() = R.layout.activity_icons

    override fun getViewModelClazz() = IconsViewModel::class.java


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        btnSubmit = findViewById(R.id.button_submit)
        btnSubmit.setOnClickListener {
            setResult(selectedIcon)
        }

        btnClear = findViewById(R.id.button_clear_icon)
        btnClear.setOnClickListener {
            setResult(null)
        }

        // список иконок
        iconsListView = findViewById(R.id.list_view_icons)
        initIconsList()
        iconsListView.onItemClickListener =
            AdapterView.OnItemClickListener { _, view: View?, i: Int, _ ->
                (iconsAdapter.getItem(i) as? TetroidIcon)?.let { icon ->
                    onIconSelected(icon, view)
                }
            }

        // список каталогов
        spFolders = findViewById(R.id.spinner_folders)

        spFolders.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            @RequiresApi(api = Build.VERSION_CODES.N)
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                foldersAdapter.getItem(position)?.let { folder ->
                    onFolderSelected(folder)
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        // загрузка
        val nodeId = intent.extras?.getString(Constants.EXTRA_NODE_ID).orEmpty()
        val iconPath = intent.extras?.getString(Constants.EXTRA_NODE_ICON_PATH).orEmpty()
        viewModel.init(nodeId, iconPath)
    }

    override fun createDependencyScope() {
        scopeSource = ScopeSource.current
    }

    override fun onBaseEvent(event: BaseEvent) {
        when (event) {
            is IconsEvent.IconsFolders -> {
                loadIconsFolders(event.folders)
            }
            is IconsEvent.CurrentIcon -> {
                selectIcon(event.icon)
            }
            is IconsEvent.LoadIconsFromFolder.InProcess -> {
                showProgress()
            }
            is IconsEvent.LoadIconsFromFolder.Success -> {
                hideProgress()
                loadIconsFromFolder(event.folder, event.icons)
            }
            else -> super.onBaseEvent(event)
        }
    }

    private fun loadIconsFolders(folders: List<String>) {
        foldersAdapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_item,
            folders
        )
        foldersAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spFolders.adapter = foldersAdapter

        if (folders.isEmpty()) {
            viewModel.logWarning(R.string.log_icons_dirs_absent)
        }
    }

    private fun loadIconsFromFolder(folder: String, icons: List<TetroidIcon>?) {
        if (icons != null) {
            // пришлось заново пересоздавать адаптер, т.к. при простой установке dataSet
            // с помощью reset(icons)
            // не сбрасывалось выделение от предыдущего списка icons (из другого раздела)
            initIconsList()
            if (selectedIcon?.folder == folder) {
                val selectedIconName = selectedIcon?.name
                val iconPosition = icons.indexOfFirst { it.name == selectedIconName }
                if (iconPosition != -1) {
                    // выделяем существующую иконку
                    val icon = icons[iconPosition]
                    selectedIcon = icon
                    iconsAdapter.selectedIcon = icon
                    btnSubmit.isEnabled = true
                }
                iconsAdapter.setData(icons)
                if (iconPosition != -1) {
                    iconsListView.setSelection(iconPosition)
                }
            } else {
                iconsAdapter.setData(icons)
            }
        }
    }

    private fun initIconsList() {
        iconsAdapter = IconsListAdapter(
            context = this,
            onLoadIconIfNeedCallback = { position, icon, imageView ->
                viewModel.loadIconIfNeed(
                    icon = icon,
                    onLoadedCallback = { drawable ->
                        // после загрузки устанавливаем значок у данного ImageView, но только если элемент еще виден на экране,
                        // т.к. его уже могли пролистать
                        if (position in iconsListView.firstVisiblePosition..iconsListView.lastVisiblePosition) {
                            imageView.setImageDrawable(drawable)
                        }
                    }
                )
            }
        )
        iconsListView.adapter = iconsAdapter
    }

    private fun onIconSelected(icon: TetroidIcon, view: View?) {
        selectedIcon = icon
        iconsAdapter.selectedIcon = icon
        iconsAdapter.setSelectedView(view)
        btnSubmit.isEnabled = true
    }

    private fun onFolderSelected(folder: String) {
        viewModel.getIconsFromFolder(folder)
    }

    private fun selectIcon(icon: TetroidIcon) {
        selectedIcon = icon
        val folderPosition = foldersAdapter.getPosition(icon.folder)
        if (folderPosition >= 0) {
            spFolders.setSelection(folderPosition)
        }
    }

    private fun setResult(icon: TetroidIcon?) {
        val intent = buildIntent {
            putExtra(Constants.EXTRA_NODE_ID, viewModel.nodeId)
            putExtra(Constants.EXTRA_NODE_ICON_PATH, icon?.path)
            putExtra(Constants.EXTRA_IS_DROP, icon == null)
        }
        setResult(RESULT_OK, intent)
        finish()
    }

    companion object {

        fun start(activity: Activity, node: TetroidNode, requestCode: Int) {
            val intent = Intent(activity, IconsActivity::class.java)
            intent.putExtra(Constants.EXTRA_NODE_ID, node.id)
            intent.putExtra(Constants.EXTRA_NODE_ICON_PATH, node.iconName)
            activity.startActivityForResult(intent, requestCode)
        }

    }

}